### PR TITLE
[4372] Implement the other degree UUIDs for HESA import

### PR DESCRIPTION
**Pull request by jhenderson**
### Context
https://trello.com/c/qfILVhHw/4372-l-implement-the-other-degree-uuids-for-hesa-import

### Changes proposed in this pull request
- New service object `Degrees::DfeReference` to help HESA and Apply import services determine the UUID for degree subject, insitution, type and grade
- Update `Degrees::CreateFromHesa` and `Degrees::MapFromApply` to use `Degrees::DfeReference`
- Remove `Hesa::Degrees` codesets in favour of the new `dfe-reference-data` gem

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

**Comment by jhenderson** [comment]
[[L] Implement the other degree UUIDs for HESA import](https://trello.com/c/qfILVhHw/4372-l-implement-the-other-degree-uuids-for-hesa-import)

**Review by markhallen — APPROVED** [review]
Looks good.

[app/services/degrees/create_from_hesa.rb](https://github.com/DFE-Digital/register-trainee-teachers/blob/10b2bae419dd90cf33a5a4a0217ce2ac85d52e20/app/services/degrees/create_from_hesa.rb) seems to have grown over time and is hard to follow. It could be useful to refactor in a future PR.

File: app/services/degrees/dfe_reference.rb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      end
+
+      def find_grade(uuid: nil, name: nil, hesa_code: nil)
+        find_item(:grades, id: uuid, hesa_code: sanitised_hesa(hesa_code), name: name)
+      end
+
+      def find_item(list_type, filters)
+        DfE::ReferenceData::Degrees.const_get(list_type.to_s.upcase).all.find do |record|
+          filters.compact.any? do |field, value|
+            same_string?(record[field], value) || almost_identical?(record[field], value)
```
Comment (gpeng) [inline app/services/degrees/dfe_reference.rb:30]: This is great moving all of this out into here but I'm still not sure about the almost identical bit. I still think `starts_with` could cause problems.

Comment (jhenderson) [reply app/services/degrees/dfe_reference.rb:30]: The "almost identical" bit came from previous code. I think it was done to deal values that might be similar at the start of the string, but different towards the end - extra information in parentheses for example. I'm happy to remove it if we think it's now safe to do so.
